pkg/server/plugin/keymanager/awskms/awskms.go

```diff
@@ -0,0 +1,533 @@
+		}
+		hashAlgo = opts.PssOptions.HashAlgorithm
+		isPSS = true
+		// opts.PssOptions.SaltLength is handled by KMS. The salt length matches the bits of the hashing algorithm.
+	default:
+		return "", kmsErr.New("unsupported signer opts type %T", opts)
+	}
+
+	isRSA := keyType == keymanager.KeyType_RSA_2048 || keyType == keymanager.KeyType_RSA_4096
+
+	switch {
+	case hashAlgo == keymanager.HashAlgorithm_UNSPECIFIED_HASH_ALGORITHM:
+		return "", kmsErr.New("hash algorithm is required")
+	case keyType == keymanager.KeyType_EC_P256 && hashAlgo == keymanager.HashAlgorithm_SHA256:
+		return types.SigningAlgorithmSpecEcdsaSha256, nil
+	case keyType == keymanager.KeyType_EC_P384 && hashAlgo == keymanager.HashAlgorithm_SHA384:
+		return types.SigningAlgorithmSpecEcdsaSha384, nil
+	case isRSA && !isPSS && hashAlgo == keymanager.HashAlgorithm_SHA256:
+		return types.SigningAlgorithmSpecRsassaPkcs1V15Sha256, nil
+	case isRSA && !isPSS && hashAlgo == keymanager.HashAlgorithm_SHA384:
+		return types.SigningAlgorithmSpecRsassaPkcs1V15Sha384, nil
+	case isRSA && !isPSS && hashAlgo == keymanager.HashAlgorithm_SHA512:
+		return types.SigningAlgorithmSpecRsassaPkcs1V15Sha512, nil
+	case isRSA && isPSS && hashAlgo == keymanager.HashAlgorithm_SHA256:
+		return types.SigningAlgorithmSpecRsassaPssSha256, nil
+	case isRSA && isPSS && hashAlgo == keymanager.HashAlgorithm_SHA384:
+		return types.SigningAlgorithmSpecRsassaPssSha384, nil
+	case isRSA && isPSS && hashAlgo == keymanager.HashAlgorithm_SHA512:
+		return types.SigningAlgorithmSpecRsassaPssSha512, nil
+	default:
+		return "", kmsErr.New("unsupported combination of keytype: %v and hashing algorithm: %v", keyType, hashAlgo)
+	}
+}
+
+func keyTypeFromKeySpec(keySpec types.CustomerMasterKeySpec) (keymanager.KeyType, error) {
+	switch keySpec {
+	case types.CustomerMasterKeySpecRsa2048:
+		return keymanager.KeyType_RSA_2048, nil
+	case types.CustomerMasterKeySpecRsa4096:
+		return keymanager.KeyType_RSA_4096, nil
+	case types.CustomerMasterKeySpecEccNistP256:
+		return keymanager.KeyType_EC_P256, nil
+	case types.CustomerMasterKeySpecEccNistP384:
+		return keymanager.KeyType_EC_P384, nil
+	default:
+		return keymanager.KeyType_UNSPECIFIED_KEY_TYPE, fmt.Errorf("unsupported key spec: %v", keySpec)
+	}
+}
+
+func keySpecFromKeyType(keyType keymanager.KeyType) (types.CustomerMasterKeySpec, error) {
+	switch keyType {
+	case keymanager.KeyType_RSA_1024:
+		return "", kmsErr.New("unsupported key type: KeyType_RSA_1024")
+	case keymanager.KeyType_RSA_2048:
+		return types.CustomerMasterKeySpecRsa2048, nil
+	case keymanager.KeyType_RSA_4096:
+		return types.CustomerMasterKeySpecRsa4096, nil
+	case keymanager.KeyType_EC_P256:
+		return types.CustomerMasterKeySpecEccNistP256, nil
+	case keymanager.KeyType_EC_P384:
+		return types.CustomerMasterKeySpecEccNistP384, nil
+	default:
+		return "", kmsErr.New("unknown key type")
+	}
+}
+
+func clonePublicKey(publicKey *keymanager.PublicKey) *keymanager.PublicKey {
```

Comment:
This probably came over from the keymanagerbase? It was an overkill step in the keymanagerbase because that package wasn't technically implementing a gRPC service. In your case, you are returning these objects from a gRPC handler and it won't mutate them. Considering the low amount of traffic for key creation/listing, you'd be fine if you left it, but it is a little confusing.